[Debug] SHOULD NEVER BE REACHED on `fast/multicol/spanner-get-re-added-on-move-crash.html`

obyknovenius opened the pull request:
#### 99535733074eea67b373aab0a73ac8d943020041
<pre>
[Debug] SHOULD NEVER BE REACHED on `fast/multicol/spanner-get-re-added-on-move-crash.html`
<a href="https://bugs.webkit.org/show_bug.cgi?id=280747">https://bugs.webkit.org/show_bug.cgi?id=280747</a>

Reviewed by Darin Adler.

It happens when the table has captions but no sections,
i.e. `sectionMoved` is `true` and `sectionCount` is `0`.

* LayoutTests/TestExpectations:
* Source/WebCore/rendering/RenderTable.cpp:
(WebCore::RenderTable::layout):

Canonical link: <a href="https://commits.webkit.org/284831@main">https://commits.webkit.org/284831@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4bf5d05f51119cf6a2ab2ee842c5222bd8acb24e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/70645 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/50051 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/23410 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/74734 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/21823 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/72762 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/57851 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/21663 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/55942 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/14415 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/73711 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/45507 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/60900 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/36404 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/42164 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/18336 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/20184 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/64098 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/18691 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/76454 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/14871 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/17898 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/63672 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/14914 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/60968 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/63613 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15641 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/11666 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/5314 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/45853 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/622 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/46927 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/48204 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/46669 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/rendering/RenderTable.cpp:
```diff
@@ -620,14 +620,12 @@ void RenderTable::layout()
-            layoutIfNeeded();
-        } else
-            ASSERT_NOT_REACHED();
+        ASSERT(m_recursiveSectionMovedWithPaginationLevel < sectionCount);
```
The old code was safer because if this assertion was not met, the code would be skipped. Was there a compelling reason to remove the `if` statement?

obyknovenius replied:
I thought the code is identical but you're right it would be different for release builds.

obyknovenius committed:
[Debug] SHOULD NEVER BE REACHED on `fast/multicol/spanner-get-re-added-on-move-crash.html`
https://bugs.webkit.org/show_bug.cgi?id=280747

Reviewed by Darin Adler.

It happens when the table has captions but no sections,
i.e. `sectionMoved` is `true` and `sectionCount` is `0`.

* LayoutTests/TestExpectations:
* Source/WebCore/rendering/RenderTable.cpp:
(WebCore::RenderTable::layout):

Canonical link: https://commits.webkit.org/284831@main
